### PR TITLE
#2782: Collapse server ACLs update events

### PR DESCRIPTION
**Pull request by mnaturel**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [ ] Technical
- [x] Other : Enhancement

## Content

Merge of multiple successive received server ACLs updates in a room into a single collapsable/expandable item.

## Motivation and context

See #2782 

## Screenshots / GIFs

<image src="https://user-images.githubusercontent.com/46314705/154494015-91455bbc-f160-48ab-9adf-6d2b4bfb5007.png" width=300><image src="https://user-images.githubusercontent.com/46314705/154494047-cd6a8212-fbec-4043-abfe-0bc1bf541274.png" width=300>

<image src="https://user-images.githubusercontent.com/46314705/154496905-80147a28-0bae-4e32-9755-933188d0a416.png" width=300><image src="https://user-images.githubusercontent.com/46314705/154496929-6e38d7f6-d150-44ad-bf29-4e16e1c72326.png" width=300>

## Tests

- Go to a room where there are many successive ACLs events
- Check the events are merged together
- Click on "expand" button
- Check all corresponding events are listed
- Click on "collapse" button
- Check all events are merged together

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s): Android 11

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MergedHeaderItemFactory.kt [hunk 1]
```diff
@@ -82,7 +83,7 @@ class MergedHeaderItemFactory @Inject constructor(private val activeSessionHolde
                                                    event: TimelineEvent,
                                                    eventIdToHighlight: String?,
                                                    requestModelBuild: () -> Unit,
-                                                   callback: TimelineEventController.Callback?): MergedMembershipEventsItem_? {
+                                                   callback: TimelineEventController.Callback?): DefaultMergedEventsItem_? {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MergedHeaderItemFactory.kt:86]: OK for the renaming, but maybe something like `SimilarEventsMergedItem` would be more appropriate. `Default` prefix is reserved for default implementation of interfaces.

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/TimelineDisplayableEvents.kt [hunk 2]
```diff
@@ -56,7 +56,8 @@ object TimelineDisplayableEvents {
 }
 
 fun TimelineEvent.canBeMerged(): Boolean {
-    return root.getClearType() == EventType.STATE_ROOM_MEMBER
+    return root.getClearType() == EventType.STATE_ROOM_MEMBER ||
+            root.getClearType() == EventType.STATE_ROOM_SERVER_ACL
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/TimelineDisplayableEvents.kt:60]: I am wondering if room member events could be merged with ACL events, which would be wrong, but your screenshots seems to show that this is not happening. Perfect!

**Review by bmarty — APPROVED** [review]
Thanks, this PR will be really appreciated once in production.
Just a remark about a class name, else LGTM!

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/TimelineDisplayableEvents.kt (hunk 2 above)
Comment (mnaturel) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/TimelineDisplayableEvents.kt:60]: Yes, it is not possible. For what I understood, in the code, there is a grouping mechanism by same event type and same day.

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MergedHeaderItemFactory.kt (hunk 1 above)
Comment (mnaturel) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MergedHeaderItemFactory.kt]: Okay, I renamed the class to `MergedSimilarEventsItem` to be consistent with existing `MergedRoomCreationItem`. Is that okay?

**Review by bmarty — APPROVED** [review]
Thanks for the update!

**Pull request merged**